client/proto/state_snapshot/v1/state_snapshot.proto

```diff
@@ -0,0 +1,159 @@
+// Copyright 2021 VMware, all rights reserved
+//
+// Concord Client's StateSnapshot Service
+
+syntax = "proto3";
+package vmware.concord.client.statesnapshot.v1;
+
+option java_package = "com.vmware.concord.client.statesnapshot.v1";
+
+// The StateSnapshot service can be used to initialize an Application from a state
+// snapshot visible to Concord Client. A state snapshot contains all key-values
+// in the `STORAGE_SPACE_MERKLE` storage space from the start of the blockchain up
+// to a specific point and is identified by a snapshot ID.
+//
+// A state snapshot contains only the last writes (as of the state snapshot itself)
+// to key-values in the `STORAGE_SPACE_MERKLE` storage space and no prior historical
+// writes.
+//
+// See `execution_engine.proto` for more information about storage spaces.
+service StateSnapshot {
+  // Get the ID of a recent and available state snapshot.
+  // Errors:
+  // UNAVAILABLE: if Concord Client is not ready yet to process requests.
+  // UNKNOWN: exact cause is unknown.
+  rpc GetRecentSnapshot(GetRecentSnapshotRequest) returns (GetRecentSnapshotResponse);
+
+  // Stream a specific state snapshot in a resumable fashion as a finite stream of key-values.
+  // Key-values are streamed with lexicographic order on keys.
+  // Errors:
+  // NOT_FOUND: if a state snapshot with the requested ID is not (or no longer) available.
+  //            Applications are advised to retry initialization by fetching a new snapshot ID.
+  // INVALID_ARGUMENT: if resuming a stream using `from_key` and that key is not part of the
+  //                   state snapshot.
+  // UNAVAILABLE: if Concord Client is not ready yet to process requests.
+  // UNKNOWN: exact cause is unknown.
+  rpc StreamSnapshot(StreamSnapshotRequest) returns (stream StreamSnapshotResponse);
+
+  // Read the values of the given keys as of a specific state snapshot.
+  // Errors:
+  // NOT_FOUND: if the state snapshot with the requested ID is not (or no longer) available.
+  //            Applications are advised to retry initialization by fetching a new snapshot ID.
+  // UNAVAILABLE: if Concord Client is not ready yet to process requests.
+  // UNKNOWN: exact cause is unknown.
+  rpc ReadAsOf(ReadAsOfRequest) returns (ReadAsOfResponse);
+}
+
+message GetRecentSnapshotRequest {
+}
+
+message GetRecentSnapshotResponse {
+  // Mandatory field.
```

Comment:
Because we're using the `optional` keyword & feature in proto3 it doesn't make sense to explicitly state what is mandatory or optional (these semantics are self-explanatory by the field declarations themselves).
To improve readability, I'd strongly suggest removing all references to optionality from the commentary as they're redundant.

Reply:
Agreed. Will change accordingly.